### PR TITLE
Adding nulling and explicit columns filter to outputs

### PR DESCRIPTION
**Pull request by amitschang**
In cases where an entire column has no non-null values, the mapping will be missing from the json entirely and as such column will be missing. Since the database still expects values we need to provide these null-filled. Being more explicit here, while a little clunky, has the advantage that we can fail fast when required columns are missing.  The view of required and nullable from the DB perspective is not necessarily the same as for the json data, optional here means the entire dataset could be missing the column.

Also adding deduplication to the dyads table, which has been seen to have redundant entries for model for example.

### CODE REVIEW

**Review by glemson — CHANGES_REQUESTED** [review]

File: susdingest/jsonloader.py [hunk 1]
```diff
@@ -56,27 +56,37 @@ def normalize_col(self, column, explode=True):
...
-                        .rename(columns=lambda x: f'journal_{x}')
-        return publications.join(citescore).convert_dtypes()
+        required_cols = ['doi', 'publication_title']
+        optional_cols = ['publication_type', 'publication_year', 'publication_month', 'citation_count',
+                         'field_weighted_citation_impact', 'journal_publishername', 'joiurnal_title',
+                         'journal_scopus_source_id', 'journal_issn_isbn']
+        publications = self.with_nullable_as_required(self.data.copy(), optional_cols + ['expressions'], required_cols)
+        publications['journal_issn_isbn'] = publications['journal_issn_isbn'] \
+            .apply(lambda x: '|'.join(str(i) for i in x) if x == x else x)
+        publications['tested_expressions'] = publications['expressions'] \
```
[the remark this answers is not in the document]
Comment (amitschang) [reply susdingest/jsonloader.py:75]: it is possible, it is added above as an optional col.

File: susdingest/jsonloader.py [hunk 2]
```diff
@@ -56,27 +56,37 @@ def normalize_col(self, column, explode=True):
...
+            .apply(lambda x: json.dumps(x) if x == x else x)
+        publications = publications.drop(columns=['expressions'])
+        if 'journal_citescore' in self.data:
+            citescore = self.normalize_col('journal_citescore', explode=False) \
+                            .rename(columns=lambda x: f'journal_{x}')
+            publications = publications.join(citescore)
+        return self.with_nullable_as_required(
+            publications, required_cols, optional_cols + ['journal_citescore_year', 'journal_citescore_value']) \
```
[the remark this answers is not in the document]
Comment (amitschang) [reply susdingest/jsonloader.py:83]: quite right. I suppose I need to add a test to excersize all optional columns - would have caught this. Thanks!

File: susdingest/jsonloader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (glemson) [reply susdingest/jsonloader.py:75]: ah yes, missed that statement. but should the column name not be 'tested_expressions'

**Comment by amitschang** [comment]
I've updated and added tests for the optional/required columns. Shouldn't have skimped in the first place since it brought out an issue in null checks, which don't work the same between pd.NA and np.nan null types, so added a function to coalesce those checks.

**Review by glemson — APPROVED** [review]
go for it

**Pull request merged**